internal/config/config_test.go

```diff
@@ -0,0 +1,218 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2025 Datadog, Inc.
+
+package config
+
+import (
+	"sync"
+	"sync/atomic"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestGet(t *testing.T) {
+	t.Run("returns non-nil config", func(t *testing.T) {
+		resetGlobalState()
+		defer resetGlobalState()
+
+		cfg := Get()
+		assert.NotNil(t, cfg, "Get() should never return nil")
+	})
+
+	t.Run("singleton behavior - returns same instance", func(t *testing.T) {
+		resetGlobalState()
+		defer resetGlobalState()
+
+		cfg1 := Get()
+		cfg2 := Get()
+		cfg3 := Get()
+
+		// All calls should return the same instance
+		assert.Same(t, cfg1, cfg2, "First and second Get() calls should return the same instance")
+		assert.Same(t, cfg1, cfg3, "First and third Get() calls should return the same instance")
+	})
+
+	t.Run("fresh config flag forces new instance", func(t *testing.T) {
+		resetGlobalState()
+		defer resetGlobalState()
+
+		// Get the first instance
+		cfg1 := Get()
+		require.NotNil(t, cfg1)
+
+		// Enable fresh config
+		SetUseFreshConfig(true)
+
+		// Get should now return a new instance
+		cfg2 := Get()
+		require.NotNil(t, cfg2)
+		assert.NotSame(t, cfg1, cfg2, "With useFreshConfig=true, Get() should return a new instance")
+
+		// Another call with useFreshConfig still true should return another new instance
+		cfg3 := Get()
+		require.NotNil(t, cfg3)
+		assert.NotSame(t, cfg2, cfg3, "With useFreshConfig=true, each Get() call should return a new instance")
+
+		// Disable fresh config
+		SetUseFreshConfig(false)
+
+		// Now it should cache again
+		cfg4 := Get()
+		cfg5 := Get()
+		assert.Same(t, cfg4, cfg5, "With useFreshConfig=false, Get() should cache the same instance")
+	})
+
+	t.Run("concurrent access is safe", func(t *testing.T) {
+		resetGlobalState()
+		defer resetGlobalState()
+
+		const numGoroutines = 100
+		var wg sync.WaitGroup
+		wg.Add(numGoroutines)
+
+		// All goroutines should get a non-nil config
+		configs := make([]*Config, numGoroutines)
+
+		for i := range numGoroutines {
+			go func(j int) {
+				defer wg.Done()
+				configs[j] = Get()
+			}(i)
+		}
+
+		wg.Wait()
+
+		// All configs should be non-nil
+		for i, cfg := range configs {
+			assert.NotNil(t, cfg, "Config at index %d should not be nil", i)
+		}
+
+		// All configs should be the same instance (singleton)
+		firstConfig := configs[0]
+		for i, cfg := range configs[1:] {
+			assert.Same(t, firstConfig, cfg, "Config at index %d should be the same instance", i+1)
+		}
+	})
+
+	t.Run("concurrent access with fresh config", func(t *testing.T) {
+		resetGlobalState()
+		defer resetGlobalState()
+
+		SetUseFreshConfig(true)
+		defer SetUseFreshConfig(false)
```

Comment:
Is there a reason why we need both `defer resetGlobalState()` and `defer SetUseFreshConfig(false)`? Both set `useFreshConfig` to be false. Could we do something like

```
resetGlobalState()
SetUseFreshConfig(true)
defer resetGlobalState()
```

instead?

Reply:
> are there use cases where we want to reset useFreshConfig but not instance

useFreshConfig=true allows to, within a given test, create new instances as many times as we like (every call to `Get()` returns a new instance)
resetGlobalState() allows us to test the real singleton behavior within a given test: it resets the sync.Once and the instance, but keeps useFreshConfig=false, therefore within a given test, only the first call to `Get()` creates a new instance and all subsequent calls return the same instance.